Add design note: Radius Copilot App integration exception scenarios (#12494)
radius-project/radius: eng/design-notes/github/2026-07-radius-copilot-app-exception-scenarios.md

# Feature Specification: Radius Copilot App Integration Exception Scenarios

- **Author**: Zach Casper (@zachcasper)

## Summary

This document describes the error and exception scenarios for the Radius integration in the GitHub Copilot app, the user experience when each occurs, and how the user recovers or corrects the problem. It is scoped to the MVP (the initial public preview) of the integration.

It is organized around the parts of the end-to-end user flow already defined in the companion document, [Radius Integration in GitHub Copilot App](radius-copilot-app-integration.md). For each part, this document enumerates the possible errors and exceptions, what the user sees, and the path back to a good state.

> The integration has three components, defined in the terms below: the **Radius skill** (in the Copilot chat), the **Radius side panel** (in the sidebar), and **Repo Radius** (which runs each operation as a GitHub Actions workflow). Where a failure originates in Repo Radius, this document references the Repo Radius outcome the skill or side panel surfaces rather than redefining it.

Author: Removed broken link and noted the document is stored outside this repo.

- **Repo Radius**: The component that runs each Radius operation as a GitHub Actions workflow and returns results as artifacts (for example, `run-rad-commands-result`, `verify-cloud-auth-result`). Defined in the [Repo Radius feature specification](https://github.com/radius-project/radius/pull/12078).
- **Assembly analysis**: The analysis the Radius skill runs on the repository to identify the application and its resources and produce the application definition.
- **Application definition**: The `app.bicep` file that describes the application, written into the `.radius` directory.
- **Application graph**: The graph of the application's resources and their relationships, shown in one of three views:
  - **Modeled application graph view**: Generated by compiling `app.bicep`; shows the application as defined, before deployment.
  - **Deployed application graph view**: Shows the resources actually deployed to the target environment.
  - **Planned application graph view**: A preview (what-if) of changes before deploying. Deferred for the MVP.

> ## Exceptions handling principles

Author: Corrected.

### Exception 0.1: Plugin install fails

The user has found the Radius plugin on the Plugins settings page and chosen to add it, but the install does not run to completion, usually because of a transient network or download error. The plugin remains listed but is not installed.

- **What the user sees:** The install does not complete and the settings page reports the failure.
- **Recovery:** Handled by the Copilot app. The user retries the install; a transient error usually clears on retry.

### Exception 0.2: Update fails

The Radius plugin is already installed and working, and the user starts an update to a newer version, but the update does not run to completion. The previously installed version is still in place.

- **What the user sees:** The update does not complete and the settings page reports the failure.
- **Recovery:** Handled by the Copilot app. The user retries the update later; the existing version keeps working in the meantime.

## Part 1: Clone the repo, invoke the Radius skill, and open the Radius side panel

> Before any Assembly analysis can begin, the repository must be brought into the session and checked out as a worktree, the Radius skill must fire and enable the repository for Radius, and the Radius side panel must come up in the sidebar. Bringing the repository in is a Copilot app prerequisite; once it is available, the first thing the skill does is enable the repository by committing the Repo Radius workflow files into `.github/workflows`, because Repo Radius runs every operation as a GitHub Actions workflow that must be committed before it can run. The exceptions here are the repository not being available (1.1), the skill not being invoked (1.2), enablement failing for lack of write permission (1.3), the default branch being protected (1.4), and the Radius side panel not coming up (1.5). All are pre-flight, so nothing has been analyzed or deployed yet, and the user recovers by correcting the cause and re-invoking the skill with no cleanup.

Reviewer: > the first thing the skill does is enable the repository by committing the Repo Radius workflow files into .github/workflows, because Repo Radius runs every operation as a GitHub Actions workflow that must be committed before it can run

As discussed, this happens as part of the deployment step. Can you update the wording to reflect that so the correct flow is captured here? Or remove it from here if this document isn't intended to cover that discussion.

Author: Updated to "Repo Radius runs every operation as a GitHub Actions workflow that must be committed before it can run; that commit happens as part of the deployment step."

- **What the user sees:** If no repository is attached, Copilot responds in chat: "Which configured repository should I show the app graph for?" If a repository is attached but cannot be cloned or checked out, the Copilot app reports that failure.
- **Recovery:** The user tells Copilot which repository to operate on.

### Exception 1.2: Skill not invoked from a natural-language prompt

The session is ready with a repository attached, and the user asks Copilot something that should invoke the Radius skill, such as asking to see the application graph. Copilot answers generally and never invokes the Radius skill, so none of the Radius flow runs and the Radius side panel never appears.

- **What the user sees:** A generic Copilot chat response with no Radius side panel and no application graph.

> - **Recovery:** The user invokes the skill explicitly with the `/radius` command (named after the plugin the user enabled), which is the deterministic entry point.

Reviewer: Looks like all the changes from the reviewed doc didn't make here. Can you update it with the addressed comments from the doc?

```suggestion
- **Recovery:** The user invokes the skill explicitly with the `/app-modeling` command (named after the plugin the user enabled), which is the deterministic entry point.
```

Author: Changed all prompts to `/app-modeling`.

- **What the user sees:** Copilot responds in chat: "To set this repository up I need to add a couple of workflow files, but I don't have write access to <repository> in this session. If you fork the repository or get write access and add it here, I can pick up where we left off."
- **Recovery:** The user forks the repository or gets write access, adds it to the session, and re-invokes the Radius skill.

### Exception 1.4: Default branch is protected

The user has write access, but the default branch (typically `main`) has branch protection that forbids committing directly to it. The skill cannot push the workflow files straight to the default branch, so instead of failing it opens a pull request that adds the Repo Radius workflow files and asks the user to merge it. The repository is not enabled until that pull request merges, because Repo Radius cannot run workflows that are not yet on the default branch.

> - **What the user sees:** Copilot responds in chat: "The default branch of <repository> is protected, so I couldn't commit the workflow files directly. I opened a pull request that adds them: <pull request link>. Merge it and I'll pick up where we left off."

Reviewer: As discussed, the skill will never push to main by default. It will commit the changes to the working tree branch and push it to the remote, and ask the user if they want a PR against main/default branch.

Author: I did some digging in the ai-extension branch. Workflow files are committed directly to main as part of the environment setup. So I collapsed 1.4 and 1.3 into one: "Cannot commit the enablement workflow files to the default branch". If it cannot commit directly, the user is offered to create a PR.

### Exception 1.5: Radius side panel does not open in the sidebar

The Radius skill runs and responds in chat, but the Radius side panel never appears in the sidebar. The plugin installs the skill and the Radius Canvas extension together, so the extension is present but failed to load (a crash on startup, a user modification, or an incompatible or corrupted build), and cannot build the side panel. The skill must detect the unavailable extension and report it clearly rather than appearing broken.

- **What the user sees:** The sidebar shows no Radius side panel, or the panel shows a load error, and Copilot responds in chat: "The Radius plugin didn't load completely, so I can't show its view in the sidebar. Reinstalling usually fixes this: remove the Radius plugin from the Plugins settings page, add it again, then restart the Copilot app."
- **Recovery:** The user restarts the Copilot app to give the extension another chance to load. If it still does not appear, they re-enable it (if disabled) or reinstall the Radius plugin for a clean build of both the skill and the extension.

> **Implementation Note:** Ideally, the user does can simply reload the extension. However, there does not seem to be a way to do this today in the Copilot app. If there is a technical solution for enabling the user to simple reload the extension, rather than reinstalling, that is highly preferred.

Author: Added.

- **Recovery:** The user adds a Dockerfile for their workload and asks Copilot to analyze the repository again.

### Exception 2.2: No recognizable application

As Copilot analyzes the repository, it cannot identify an application or its application resources. Nothing is written.

- **What the user sees:** Copilot responds in chat: "I looked through the repository but could not identify an application or application resources. Which directory contains your application source code and Dockerfile?"
- **Recovery:** The user responds with a directory name in the repository containing identifiable application resources, then asks Copilot to analyze the repository again.

### Exception 2.3: An application definition already exists

The repository already contains a `.radius` directory with an `app.bicep` application definition from a previous Assembly analysis, and the user asks Copilot to analyze the repository again. Rather than silently overwriting the existing `app.bicep`, Copilot stops and asks how to proceed.

> - **What the user sees:** Copilot responds in chat: "This repository already has an application definition in the `.radius` directory. Would you like me to keep it, or analyze the repository again and overwrite `app.bicep`?"

Reviewer: As discussed, if the prompt is initiated from a new Copilot session, we publish app.bicep (and other artifacts) to a new remote branch in the repository that matches the user's working tree branch. We then prompt the user to create a pull request against main. So the flow described here doesn't exactly match the user experience we've discussed and implemented.

Author: Updated part 2 overview with this flow.

- **Recovery:** The user chooses to keep the existing definition, or to regenerate it, which overwrites the existing `app.bicep`.

### Exception 2.4: Assembly analysis fails

The user starts the Assembly analysis on a repository, but it does not successfully complete, for example because of a transient model or tooling error. Nothing is written on a failed Assembly analysis.

- **What the user sees:** Copilot responds in chat: "I wasn't able to finish analyzing the repository this time. Want me to try again?"
- **Recovery:** The user retries; the Assembly analysis is re-runnable and safe to repeat because nothing is written until it succeeds.

### Exception 2.5: Cannot merge the `.radius` files into a protected branch

Copilot works off a worktree branch in a local clone, so writing the generated files to that branch always succeeds and the `.radius` directory is created there. The exception arises when those files need to land on the default branch (typically `main`) and that branch is protected: Copilot cannot commit to it directly, so instead of failing it opens a pull request that adds the `.radius` directory and asks the user to merge it.

> - **What the user sees:** Copilot responds in chat: "I generated the `.radius` files on the worktree branch, but the default branch of <repository> is protected, so I couldn't commit them there directly. I opened a pull request that adds them: <pull request link>. Merge it and the application definition will be on <branch>."

Reviewer: This is the currently agreed upon user experience. The only time it will fail is if the worktree branch on remote is protected.

```suggestion
- **What the user sees:** Copilot responds in chat: "I generated the `.radius` files on the worktree branch. Would you like me to open a pull request that adds them: <pull request link> to the default branch of <repository>."
```

Author: App assembly always writes the app.bicep to the local worktree branch. It doesn't commit or push. That's left to the user. So 2.5 is really just when the skill cannot write the `.radius` files to the worktree (kind of an edge case).

I've reframed 2.5 to this edge case, but added 5.5 Application definition not found on the deploy branch.

- **Recovery:** The user merges the pull request that adds the `.radius` directory into the default branch.

### Exception 2.6: Generated `app.bicep` is invalid

After the Assembly analysis writes the `.radius` directory, Copilot compiles `app.bicep` to generate the modeled application graph view. If the generated `app.bicep` does not compile, the view cannot be generated, and Copilot reports the compile error instead of showing the graph.

> - **What the user sees:** Nothing. Copilot identifies that the app.bicep file is not well formed and continues to modify it until the compilation completes successfully.

Author: Ignoring this suggestion. As-is is correct.

### Exception 3.1: Modeled application graph view missing or unreadable

The Assembly analysis completed successfully, so `app.bicep` and the modeled application graph view were both written to the `.radius` directory. But when the user opens the Radius side panel, the modeled application graph view does not appear, because it is missing or cannot be read, so the side panel has nothing to display.

- **What the user sees:** Copilot responds in chat: "I cannot find the modeled application graph view to show. I can rebuild it from the application definition then show it."
- **Recovery:** Copilot automatically regenerates the modeled application graph view from the `app.bicep` file and renders the view in the Radius side panel.

### Exception 3.2: Empty application graph

The user opens the Radius side panel and the modeled application graph view loads, but it contains no resources, so the side panel renders an empty graph.

- **What the user sees:** The Radius side panel renders but shows no resources on the graph. There are no error messages or buttons for the user to click.

> - **Recovery:** The asks Copilot to troubleshoot why there are no resources on the application graph and/or asks Copilot to perform the Assembly analysis again and recreate the application definition and application graph.

Author: Added
